File: TDD/TDD Coding Practice Arithmetic/Double Stack Algorithm/Code/Java/src/fayelab/tdd/arithmetic/doublestack/pow/DoubleStackProcessor.java
package fayelab.tdd.arithmetic.doublestack.pow;

import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.function.BiFunction;
import java.util.regex.Pattern;
import java.util.stream.IntStream;
import static java.util.stream.Collectors.toList;

public class DoubleStackProcessor
{
    private static final int OPERAND_STACK_MAX_SIZE = 3;
    private static final int OPERATOR_STACK_MAX_SIZE = 2;
    
    private static Map<Character, BiFunction<Integer, Integer, Integer>> operatorAndFuncMap2= null;
        
    static
    {
        operatorAndFuncMap2 = new HashMap<>();
        operatorAndFuncMap2.put('+', (x, y) -> x + y);
        operatorAndFuncMap2.put('-', (x, y) -> x - y);
        operatorAndFuncMap2.put('*', (x, y) -> x * y);
        operatorAndFuncMap2.put('/', (x, y) -> x / y);
        operatorAndFuncMap2.put('%', (x, y) -> x % y);
    }
    
    private static Map<String, BiFunction<Integer, Integer, Integer>> operatorAndFuncMap= null;
    
    static
    {
        operatorAndFuncMap = new HashMap<>();
        operatorAndFuncMap.put("+", (x, y) -> x + y);
        operatorAndFuncMap.put("-", (x, y) -> x - y);
        operatorAndFuncMap.put("*", (x, y) -> x * y);
        operatorAndFuncMap.put("/", (x, y) -> x / y);
        operatorAndFuncMap.put("%", (x, y) -> x % y);
    }
    
    private static Map<Character, Integer> operatorAndPriorityMap2 = null;
    
    static
    {
        operatorAndPriorityMap2 = new HashMap<>();
        operatorAndPriorityMap2.put('+', 1);
        operatorAndPriorityMap2.put('-', 1);
        operatorAndPriorityMap2.put('*', 2);
        operatorAndPriorityMap2.put('/', 2);
        operatorAndPriorityMap2.put('%', 2);
    }
    
    private static Map<String, Integer> operatorAndPriorityMap = null;
    
    static
    {
        operatorAndPriorityMap = new HashMap<>();
        operatorAndPriorityMap.put("+", 1);
        operatorAndPriorityMap.put("-", 1);
        operatorAndPriorityMap.put("*", 2);
        operatorAndPriorityMap.put("/", 2);
        operatorAndPriorityMap.put("%", 2);
    }
        
    private int[] operands = new int[OPERAND_STACK_MAX_SIZE];
    private int idxOfOperands = 0;
    private char[] operators2 = new char[OPERATOR_STACK_MAX_SIZE];
    private String[] operators = new String[OPERATOR_STACK_MAX_SIZE];
    private int idxOfOperators = 0;

    public void process(String item)
    {
        if(allDigits(item))
        {
            processOperand(item);
        }
        else
        {
            processOperator(item);
        }
    }

    public void process2(String item)
    {
        if(allDigits(item))
        {
            processOperand(item);
        }
        else
        {
            processOperator2(item);
        }
    }

    private void processOperand(String item)
    {
        pushOperand(toInt(item));
    }

    private void processOperator2(String item)
    {
        char curChar = toChar(item);
        
        while(notEmptyOperatorStack() && notPriorTo2(curChar, topOperator2()))
        {
            calcOnce2();
        }
        
        pushOperator2(curChar);
    }

    private void processOperator(String item)
    {
        while(notEmptyOperatorStack() && notPriorTo(item, topOperator()))
        {
            calcOnce();
        }
        
        pushOperator(item);
    }

    public int result()
    {
        calc();
        return popOperand();
    }

    private void calc()
    {
        while(notEmptyOperatorStack())
        {
            calcOnce2();
        }
    }
    
    private boolean notPriorTo(String operator1, String operator2)
    {
        return priority(operator1) <= priority(operator2);
    }

    private boolean notPriorTo2(char operator1, char operator2)
    {
        return priority2(operator1) <= priority2(operator2);
    }
    
    private int priority(String operator)
    {
        return operatorAndPriorityMap.get(operator);
    }

    private int priority2(char operator)
    {
        return operatorAndPriorityMap2.get(operator);
    }
    
    private void calcOnce()
    {
        int rOperand = popOperand();
        int lOperand = popOperand();
        String operator = popOperator();
        pushOperand(apply(operator, lOperand, rOperand));
    }

    private void calcOnce2()
    {
        int rOperand = popOperand();
        int lOperand = popOperand();
        char operator = popOperator2();
        pushOperand(apply2(operator, lOperand, rOperand));
    }
    
    private int apply(String operator, int lOperand, int rOperand)
    {
        return operatorAndFuncMap.get(operator).apply(lOperand, rOperand);
    }

    private int apply2(char operator, int lOperand, int rOperand)
    {
        return operatorAndFuncMap2.get(operator).apply(lOperand, rOperand);
    }

    void pushOperand(int operand)
    {
        operands[idxOfOperands++] = operand;
    }

    private int popOperand()
    {
        return operands[--idxOfOperands];
    }
    
    private void pushOperator(String operator)
    {
        operators[idxOfOperators++] = operator;
    }

    private void pushOperator2(char c)
    {
        operators2[idxOfOperators++] = c;
    }
    
    private String popOperator()
    {
        return operators[--idxOfOperators];
    }
    
    private char popOperator2()
    {
        return operators2[--idxOfOperators];
    }
    
    private String topOperator()
    {
        return operators[idxOfOperators - 1];
    }

    private char topOperator2()
    {
        return operators2[idxOfOperators - 1];
    }
    
    private boolean allDigits(String str)
    {
        return Pattern.matches("\\d+", str);
    }
    
    private boolean notEmptyOperatorStack()
    {
        return idxOfOperators != 0;
    }

    private int toInt(String str)
    {
        return Integer.parseInt(str);
    }

    private char toChar(String oneCharStr)
    {
        return oneCharStr.charAt(0);
    }

    int getOperandStackSize()
    {
        return idxOfOperands;
    }

    int getOperandFromStack(int idx)
    {
        return operands[idx];
    }

    int getOperatorStackSize()
    {
        return idxOfOperators;
    }

    char getOperatorFromStack(int idx)
    {
        return operators2[idx];
    }

    List<Integer> dumpOperandStack()
    {
        return IntStream.range(0, idxOfOperands).mapToObj(i -> operands[i]).collect(toList());
    }

    List<String> dumpOperatorStack()
    {
        return IntStream.range(0, idxOfOperators).mapToObj(i -> operators[i]).collect(toList());
    }
    
    List<Character> dumpOperatorStack2()
    {
        return IntStream.range(0, idxOfOperators).mapToObj(i -> operators2[i]).collect(toList());
    }
}

File: TDD/TDD Coding Practice Arithmetic/Double Stack Algorithm/Code/Python/negint/doublestackprocessor.py
import copy
import re

class DoubleStackProcessor(object):
    operator_func_map = {'+':lambda x, y: x + y,
                         '*':lambda x, y: x * y,
                         '-':lambda x, y: x - y,
                         '/':lambda x, y: x / y,
                         '%':lambda x, y: x % y,
                         '**':lambda x, y: x ** y}

    operator_priority_map = {'+':1, '-':1, '*':2, '/':2, '%':2, '**':3}
    
    def __init__(self):
        self.operandstack = []
        self.operatorstack = []

    def process(self, item):
        if self.isint(item):
            self.process_operand(item)
        else:
            self.process_operator(item)

    def process_operand(self, item):
        self.push_operand(int(item))

    def process_operator(self, item):
        while self.not_empty_operatorstack() and self.not_prior_to(item, self.top_operator()):
            self.calc_once()

        self.push_operator(item)

    def result(self):
        self.calc()
        return self.pop_operand()

    def calc(self):
        while self.not_empty_operatorstack():
            self.calc_once()

    def not_empty_operatorstack(self):
        return len(self.operatorstack) != 0

    def not_prior_to(self, operator1, operator2):
        return self.priority(operator1) <= self.priority(operator2)

    def priority(self, operator):
        return DoubleStackProcessor.operator_priority_map[operator]

    def calc_once(self):
        r_operand = self.pop_operand()
        l_operand = self.pop_operand()
        operator = self.pop_operator()
        self.push_operand(self.apply(operator, l_operand, r_operand))

    def apply(self, operator, l_operand, r_operand):
        return DoubleStackProcessor.operator_func_map[operator](l_operand, r_operand)

    def push_operand(self, operand):
        self.operandstack.append(operand)

    def pop_operand(self):
        return self.operandstack.pop()

    def push_operator(self, operator):
        self.operatorstack.append(operator)

    def pop_operator(self):
        return self.operatorstack.pop()

    def top_operator(self):
        return self.operatorstack[-1]

    def isint(self, item):
        return re.match(r'-?\b[0-9]+\b', item)

    def _dump_operandstack(self):
        return copy.copy(self.operandstack)

    def _dump_operatorstack(self):
        return copy.copy(self.operatorstack)
